### PR TITLE
Add Configurable Products to Cart Without Variant SKU

### PR DESCRIPTION
**Pull request by pmclain**
### Description (*)
* Add a module method for building buy requests used for adding items to cart
* Only return visible cart items
* Add configurable products to cart using the parent sku and desired configurations, matching the behavior of the existing storefront

### Fixed Issues (if relevant)
1. magento/graphql-ce#438

### Manual testing scenarios (*)
1. Create empty cart
2. Fetch configurable product information:
```graphql
{
  products(
    search:"configurable"
    pageSize:10
  ) {
    items {
      sku
      ... on ConfigurableProduct {
        configurable_options {
          attribute_id
          attribute_code
          id
          label
          position
          product_id
          use_default
          values {
            default_label
            label
            store_label
            use_default_value
            value_index
          }
        }
      }
    }
  }
}
```
3. Add configurable product to cart:
```graphql
mutation addConfigurableProductToCart(
  $cart_id:String!
  $sku:String!
) {
  addConfigurableProductsToCart(
    input:{
      cart_id:$cart_id
      cart_items:{
        configurable_attributes:[{
            id:190
            value:271
          },
        	{
            id:191
            value:273
          }
        ]
        data:{
          sku:$sku
          quantity:1
        }
      }
    }
  ) {
    cart {
      items {
        id
        quantity
        product {
          sku
        }
        ... on ConfigurableCartItem {
          configurable_options {
            id
            option_label
            value_id
            value_label
          }
        }
      }
    }
  }
}
```

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by lenaorobei** [comment]
@pmclain thank you for the contribution. Such kind of changes requires detailed architecture review.

**Comment by TomashKhamlai** [comment]
Could you please merge 2.3-develop to this branch or commit cc65d3e6fbb53de01cdd3cd36ba1b78ef0ca3bc1 and everything before it?
Cannot place order 
```graphql
{
  "errors": [
    {
      "message": "Unable to place order: Some addresses can't be used due to the configurations for specific countries.",
      "category": "graphql-input",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "placeOrder"
      ]
    }
  ],
  "data": {
    "placeOrder": null
  }
}
```

**Comment by TomashKhamlai** [comment]
Maybe it is a mistake ConfigurableProductCartItemInput.configurable_attributes: [ConfigurableCartItemAttributesInput]!

```graphql
mutation addConfigurableProductToCart(
  $cart_id: String!
  $sku: String!
) {
  addConfigurableProductsToCart(
    input: {
      cart_id: $cart_id
      cart_items: {
        configurable_attributes: [
          {
            id: 93
            value: 4
          }
          {
            id: 93
            value: 5
          }
          {
            id: 93
            value: 6
          }
          {
            id: 93
            value: 7
          }
        ]
        data: {
          sku: $sku
          quantity: 4
        }
      }
    }
  ) {
    cart {
      items {
        id
        quantity
        product {
          sku
        }
        ... on ConfigurableCartItem {
          configurable_options {
            id
            option_label
            value_id
            value_label
          }
        }
      }
    }
  }
}
```
On storefront we can select only one configuration. Your code acts the same like on storefront, but schema gives me falsy expectations.

**Comment by pmclain** [comment]
@TomashKhamlai The schema needs to support multiple configurable attributes, since a configuration may have multiple configurable attributes required ie: size and color.

@naydav should we add input validation that throws an error when multiple inputs contain the same attribute id?

**Comment by TomashKhamlai** [comment]
Please, update your branch to the latest mainline as soon as possible. I saw bugs that are already fixed.

**Comment by TomashKhamlai** [comment]
Found an issue:
1. Add one variation of Configurable Product with qty 1 to Cart from Storefront as registered Customer (attribute: color with 2 text swatches)
2. Send request:
```graphql
query viewCart(
  $cart_id: String!
) {
  cart(
    cart_id: $cart_id
  ) {
    items {
      ... on ConfigurableCartItem {
        customizable_options {
          is_required
        }
      }
    }
  }
}
```
Response:
```graphql
{
  "errors": [
    {
      "debugMessage": "Cannot return null for non-nullable field ConfigurableCartItem.customizable_options.",
      "message": "Internal server error",
      "category": "internal",
      "locations": [
        {
          "line": 5,
          "column": 9
        }
      ],
      "path": [
        "cart",
        "items",
        0,
        "customizable_options"
      ]
    }
  ],
  "data": {
    "cart": {
      "items": [
        null
      ]
    }
  }
}
```

**Additional notes:**
Magento is in production mode
Configurable Product has a customizable option of field type. But it is not required and it was not used.
When I send
```graphql
query viewCart(
  $cart_id: String!
) {
  cart(
    cart_id: $cart_id
  ) {
    items {
      ... on ConfigurableCartItem {
        quantity
      }
    }
  }
}
```
It looks fine, the response is 
```graphql
{
  "data": {
    "cart": {
      "items": [
        {
          "quantity": 1
        }
      ]
    }
  }
}
```

Maybe it will be useful to know
```graphql
{
  "data": {
    "products": {
      "items": [
        {
          "options": [
            {
              "option_id": 1,
              "required": false,
              "sort_order": 1,
              "title": "Package",
              "__typename": "CustomizableFieldOption"
            }
          ],
          "__typename": "ConfigurableProduct",
          "sku": "Configurable Product 1",
          "configurable_options": [
            {
              "attribute_id": "93",
              "attribute_code": "color",
              "id": 1,
              "label": "Color",
              "position": 0,
              "product_id": 3,
              "use_default": false,
              "values": [
                {
                  "default_label": "white",
                  "label": "white",
                  "store_label": "white",
                  "use_default_value": true,
                  "value_index": 4
                },
                {
                  "default_label": "black",
                  "label": "black",
                  "store_label": "black",
                  "use_default_value": true,
                  "value_index": 5
                }
              ]
            }
          ]
        }
      ]
    }
  }
}
```

**Comment by TomashKhamlai** [comment]
It would be awesome if you could fix the file conflict and update the code to the latest mainline today :pray:.
It will allow setting the milestone and decide whether it's going to be included in the next release.

**Comment by TomashKhamlai** [comment]
@pmclain, I can confirm that information provided here https://github.com/magento/graphql-ce/pull/699#issuecomment-515431404
is still reproducible

**Comment by pmclain** [comment]
@TomashKhamlai I'm working on the merge. I'll give you a mention when ready.

**Comment by pmclain** [comment]
@TomashKhamlai Merge complete. This is ready for QA. I added the same cart item update fix as in #396.

File: app/code/Magento/ConfigurableProductGraphQl/Model/Resolver/ConfigurableCartItemOptions.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        }
+        /** @var Item $cartItem */
+        $cartItem = $value['model'];
+
+        $result = [];
+        foreach ($this->configurationHelper->getOptions($cartItem) as $option) {
+            $result[] = [
+                'id' => $option['option_id'],
+                'option_label' => $option['label'],
+                'value_id' => $option['option_value'],
```
Comment (TomashKhamlai) [inline app/code/Magento/ConfigurableProductGraphQl/Model/Resolver/ConfigurableCartItemOptions.php:61]:
```graphql
{
  "errors": [
    {
      "debugMessage": "Notice: Undefined index: option_value in /shared/httpd/graphql-ce/graphql-ce/app/code/Magento/ConfigurableProductGraphQl/Model/Resolver/ConfigurableCartItemOptions.php on line 61",
      "message": "Internal server error",
      "category": "internal",
      "locations": [
        {
          "line": 9,
          "column": 9
        }
      ],
      "path": [
        "cart",
        "items",
        0,
        "configurable_options"
      ]
    }
  ],
  "data": {
    "cart": {
      "applied_coupon": null,
      "email": "John.Doe@example.com",
      "items": [
        null
      ],
      "prices": {
        "grand_total": {
          "value": 14.75,
          "currency": "USD"
        }
      },
      "shipping_addresses": [
        {
          "postcode": "90230"
        }
      ],
      "billing_address": {
        "postcode": "90230"
      }
    }
  }
}
```

The request was:
```graphql
query viewCart(
  $cart_id: String!
) {
  cart(
    cart_id: $cart_id
  ) {
    applied_coupon {
      code
    }
    email
    items {
      ... on ConfigurableCartItem {
        configurable_options {
          id
          option_label
          value_id
          value_label
        }
        customizable_options {
          id
          is_required
          label
          sort_order
          values {
            id
            label
            price {
              type
              units
              value
            }
            value
          }
        }
      }
      product {
        sku
        stock_status

        ... on ConfigurableProduct {
          attribute_set_id
          color
          sku
          variants {
            attributes {
              code
              label
              value_index
            }
            product {
              options {
                option_id
                required
                sort_order
                title
                __typename
              }
              tier_prices {
                customer_group_id
                percentage_value
                qty
                value
                website_id
              }
              product_links {
                link_type
                linked_product_sku
                linked_product_type
                position
                sku
                __typename
              }
            }
          }
        }

        ... on SimpleProduct {
          gift_message_available
          attribute_set_id
          categories {
            __typename
            url_path
            url_key
            product_count
            path_in_store
            level
            display_mode
            path
          }
          websites {
            code
            default_group_id
            id
            is_default
            name
          }
          weight
          type_id
          canonical_url
          url_key
          url_path
          special_price
          tier_price
          tier_prices {
            customer_group_id
            percentage_value
            qty
            value
            website_id
          }
        }
      }
      quantity
    }
    prices {
      grand_total {
        value
        currency
      }
    }
    shipping_addresses {
      postcode
    }
    billing_address {
      postcode
    }
  }
}
```

**Comment by TomashKhamlai** [comment]
Probably it is already reported. Just informing you. Also, it will help me in the future (#663)
```graphql
{
  "errors": [
    {
      "debugMessage": "Cannot return null for non-nullable field ConfigurableCartItem.customizable_options.",
      "message": "Internal server error",
      "category": "internal",
      "locations": [
        {
          "line": 7,
          "column": 9
        }
      ],
      "path": [
        "cart",
        "items",
        0,
        "customizable_options"
      ]
    }
  ],
  "data": {
    "cart": {
      "items": [
        null
      ]
    }
  }
}
```

The request was
```graphql
query viewCart (
  $cart_id: String!
) {
  cart(
    cart_id: $cart_id
  ) {
    items {
      ... on ConfigurableCartItem {
        quantity
        id
        customizable_options {
          id
          values {
            id
            value
          }
        }
        configurable_options {
          id
          value_id
        }
      }
    }
  }
}
```

**Comment by TomashKhamlai** [comment]
Before changes in c7d9130289993310992cb3e4a0b5ec7c3dd196bd we had the problem which would sound like "_The cart is not updating properly: new cart items are created instead of changing quantity while adding the same product to the cart. At the same time update of the first cart item do not trigger recalculation while update of the last one joins them back together._", but now it sounds like "_The cart is not updating properly: and update of the first item can be mysteriously rolled back to the initial state by the update of the last cart item and cart items are never joined, while they should_".

The problem sounds weird and can pollute the conversation here. I am thinking about excluding changes that are relative to mutation updateCartItem. Because it looks like a complex problem for me and can be addressed as a separate PR. @naydav hasn't approved this approach yet. But he approved that any problems related to updateCartItem can be reported as a separate issue. The main reason for it is that the problem became common for different types of products except for the simple product.

If you see any advantage of the approach that I suggested, please contact @naydav in Slack. Otherwise, feel free to continue with updateCartItem here in this PR. I will link this PR to the issues that I've described briefly.

**Comment by TomashKhamlai** [comment]
We still have some failures. Mention m in comments if we need confirmation that problems were fixed.

**Pull request merged**